### PR TITLE
Scheme creation/tree visualization

### PR DESCRIPTION
**Pull request by gcttong**
Added tree visualization feature, which displays subclade information on the phylogenetic tree

### CODE REVIEW

File: biohansel/create/schema_generator.py [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+        group: the group membership of the SNV
+        reference_snv: the genome sequence of the reference genome
+        alternate_snv: the genome sequence of the alternate form of the SNV
+
+    Returns:
+        sequence_string: the sequence string to be outputted into the schema file
+
+    """
+
+    return (textwrap.dedent(f"""\
```
Comment (peterk87) [inline biohansel/create/schema_generator.py]: Please remove the `textwrap.dedent` call and either convert this to a multiline string that doesn't have leading whitespace or into a multiple single line f-strings e.g.

```python
    return f""">({chromosome}){position}-{group}
{alternate_snv if ratio_value > 0 else reference_snv}
>negative({chromosome}){position}-{group}
{reference_snv if ratio_value > 0 else alternate_snv}
"""

# OR

    return (f">({chromosome}){position}-{group}"
            f"{alternate_snv if ratio_value > 0 else reference_snv}"
            f">negative({chromosome}){position}-{group}"
            f"{reference_snv if ratio_value > 0 else alternate_snv}")
```

File: biohansel/create/io/output.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
+import os
+
+import pandas as pd
+
+from typing import List
+
+from biohansel.create.schema_generator import get_sequence_string
+
+def write_sequence_file(output_directory: str, df_list: List[pd.DataFrame],
```
Comment (peterk87) [inline biohansel/create/io/output.py]: Please extract a function that returns a generator that yields the tile sequences as a string in FASTA format.

File: biohansel/create/io/output.py [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+    Returns:
+         Creates schema file in the output directory specified by the user
+    """
+
+    with open(os.path.join(output_directory, f"{schema_name}.fasta"),
+              "a+") as file:
+        for curr_df in df_list:
+            for index, row in curr_df.iterrows():
+                ratio_value = row.iloc[3]
```
Comment (peterk87) [inline biohansel/create/io/output.py]: What's 3? What field/column does the 3 correspond to?

File: biohansel/create/io/output.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+                position = index
```
Comment (peterk87) [inline biohansel/create/io/output.py]: `position = row['POS']`?

File: biohansel/create/schema_generator.py [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+    >({chromosome}){position}-{group}
+    {alternate_snv if ratio_value > 0 else reference_snv}
+    >negative({chromosome}){position}-{group}
+    {reference_snv if ratio_value > 0 else alternate_snv}\n"""))
+
+def group_snvs(
```
Comment (peterk87) [inline biohansel/create/schema_generator.py]: From this function, can you extract a function that takes a list of in-group genomes and a list of out-group genomes and returns the indices of SNPs that differentiate those 2 groups of genomes?

e.g.
```python
def find_snps_discriminating_2_groups(df_vcf: pd.DataFrame, ingroup_genomes: List[str], outgroup_genomes: List[str]) -> List[int]: # OR pd.Series OR pd.DataFrame (i.e. filtered VCF DF)
```

File: biohansel/create/cluster_generator.py [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+    Args:
+        cluster_genomes_df: the dataframe containing the genomes and their respective subtypes at each threshold
+
+    Returns:
+        final_cluster_dict: the dictionary that contains the final subtype assignments for each genome
+
+    """
+
+    final_cluster_dict = {}
+    for genome, row in cluster_genomes_df.apply(row_subtype, axis=1).iteritems():
```
Comment (peterk87) [inline biohansel/create/cluster_generator.py]: [`pd.Series.to_dict()`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.to_dict.html)?

File: biohansel/create/cluster_generator.py [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+
+    Returns:
+        row_unique[-1]: the last valid subtype name for that genome
+
+    """
+
+    row_unique = curr_genome.unique()
+    row_unique = row_unique[(row_unique != '') & (~pd.isnull(row_unique))]
+
+    return row_unique[-1]
```
Comment (peterk87) [inline biohansel/create/cluster_generator.py]: Are you sure the last unique item will be the final subtype for each row? It may be a good idea to sort by string length to be safe.

File: biohansel/create/cluster_generator.py [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+                    genomes
+
+    Returns:
+        modified_cluster_dict: A dictionary with values beings filled for all cells
+
+    """
+
+    modified_cluster_dict = {}
+    for threshold, groupings in cluster_dict.items():
+        threshold_dict = {}
```
Comment (peterk87) [inline biohansel/create/cluster_generator.py]: `threshold_dict` -> `genome_cluster`

Also could be dict comprehension:

```python
out[threshold] = {genome: cluster for cluster, genomes in cluster_genomes.items() for genome in genomes}
```

File: biohansel/create/cluster_generator.py [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+        cluster_dict: a dictionary that contains the cluster groupings of each genome with each value being sets of
+                    genomes
+
+    Returns:
+        modified_cluster_dict: A dictionary with values beings filled for all cells
+
+    """
+
+    modified_cluster_dict = {}
+    for threshold, groupings in cluster_dict.items():
```
Comment (peterk87) [inline biohansel/create/cluster_generator.py]: `groupings` -> `cluster_genomes`

File: biohansel/create/cluster_generator.py [hunk 10]
```diff
@@ -0,0 +1,236 @@
...
+    Args:
+        cluster_dict: a dictionary that contains the cluster groupings of each genome with each value being sets of
+                    genomes
+
+    Returns:
+        modified_cluster_dict: A dictionary with values beings filled for all cells
+
+    """
+
+    modified_cluster_dict = {}
```
Comment (peterk87) [inline biohansel/create/cluster_generator.py]: Rename `modified_cluster_dict` to `out`

File: biohansel/create/cluster_generator.py [hunk 11]
```diff
@@ -0,0 +1,236 @@
...
+        cluster_dict[grouping].append(threshold)
+
+    cluster_matrix = matrix[[thresholds[0] for thresholds in cluster_dict.values()]]
+    cluster_matrix.columns = [index for index, thresholds in enumerate(cluster_matrix.columns)]
+
+    cluster_matrix = cluster_matrix.drop([0], axis=1)
+
+    clusters_genomes_dict = cluster_df_to_dict(cluster_matrix)
+
+    hierarchical_cluster_df = pd.DataFrame(
```
Comment (peterk87) [inline biohansel/create/cluster_generator.py]: This is very dense, please create some local variables. Also it would be useful to return `hierarchical_cluster_df` so users can see how the clusters were assigned at various thresholds. 

Also, try to put the `df` before the rest of the var name to be consistent with the rest of the codebase, thanks.

File: biohansel/create/cluster_generator.py [hunk 12]
```diff
@@ -0,0 +1,236 @@
...
+    cluster_dict = defaultdict(list)
+
+    for threshold in thresholds:
+        grouping = '-'.join([str(cluster) for cluster in matrix[threshold]])
+        cluster_dict[grouping].append(threshold)
+
+    cluster_matrix = matrix[[thresholds[0] for thresholds in cluster_dict.values()]]
+    cluster_matrix.columns = [index for index, thresholds in enumerate(cluster_matrix.columns)]
+
+    cluster_matrix = cluster_matrix.drop([0], axis=1)
```
Comment (peterk87) [inline biohansel/create/cluster_generator.py]: Refactor lines 84 to 99 into a function called `distinct_flat_clusters` returning a `pd.DataFrame` with the same orientation as `cluster_df` above minus the columns producing duplicated clusters.

File: biohansel/create/cluster_generator.py [hunk 13]
```diff
@@ -0,0 +1,236 @@
...
+        distance_matrix: a matrix of pair-wise distances between samples
+        min_group_size: the minimum child group size for each new subtype branching point from the parent group
+
+    Returns:
+         flat_clusters: an array of flat clusters from the clustering array
+    """
+    clusters = np.array([
+        fcluster(clustering_array, t=distance, criterion='distance')
+        for distance in np.sort(np.unique(distance_matrix))])
+    cluster_df = pd.DataFrame(np.array(clusters), index=np.sort(np.unique(distance_matrix)))
```
Comment (peterk87) [inline biohansel/create/cluster_generator.py]: This function should return `cluster_df` here. The rest of the code in this functions should be refactored into other functions.

File: biohansel/create/cluster_generator.py [hunk 14]
```diff
@@ -0,0 +1,236 @@
...
+    Returns:
+        clustering_array: a hierarchical clustering linkage array that is calculated from the distance matrix
+    """
+
+    clustering_array = linkage(distance_matrix, method='complete')
+
+    return clustering_array
+
+
+def output_flat_clusters(clustering_array: np.ndarray, genomes_only: List, distance_matrix: np.ndarray,
```
Comment (peterk87) [inline biohansel/create/cluster_generator.py]: This function isn't returning the flat clusters at various distance threshold, instead, it is returning the hierarchical clusters. Please split this function up into its component parts so that we can output the intermediate output to show how the final HC clusters are derived.

File: biohansel/create/cluster_generator.py [hunk 15]
```diff
@@ -0,0 +1,236 @@
...
+import pandas as pd
+import scipy as sp
+
+
+from collections import defaultdict
+from scipy.cluster.hierarchy import fcluster, linkage
+from scipy.spatial.distance import pdist
+
+
+def find_clusters(df: pd.DataFrame, min_group_size: int) -> Dict[str, int]:
```
Comment (peterk87) [inline biohansel/create/cluster_generator.py]: What about user specified distance threshold levels? If `None` then you can use the unique distances from the distance matrix. This function should be returning all the intermediate outputs as well. Please look into creating an [attrs](http://www.attrs.org/en/stable/) class to store this information like the `Subtype` class.

File: biohansel/create/cluster_generator.py [hunk 16]
```diff
@@ -0,0 +1,236 @@
...
+
+    Args:
+        df: the DataFrame that contains only the samples' binary data
+
+    Returns:
+        a matrix of pair-wise distances between samples
+    """
+
+    return sp.spatial.distance.pdist(
+        df.transpose(), metric='hamming')
```
Comment (peterk87) [inline biohansel/create/cluster_generator.py]: The pairwise distance metric could be specified via command-line with `hamming` as the default. Users might want to compute `matching` or `euclidean` distances. See http://click.pocoo.org/6/options/#choice-options for implementing choice options.

File: biohansel/create/cluster_generator.py [hunk 17]
```diff
@@ -0,0 +1,236 @@
...
+
+
+def create_linkage_array(distance_matrix: np.ndarray) -> np.ndarray:
```
Comment (peterk87) [inline biohansel/create/cluster_generator.py]: Let's make the linkage method up to the user via the command-line with `complete` as the default. See http://click.pocoo.org/6/options/#choice-options

File: biohansel/cli.py [hunk 18]
```diff
@@ -276,6 +312,37 @@ def create(vcf_file_path, reference_genome_path, phylo_tree_path, distance_thres
...
+    if schema_name is not None:
+        schema_name = f"{schema_name}-{schema_version}"
+    else:
+        schema_name = f"bio_hansel-schema-{reference_genome_name}-{schema_version}"
+
+    if not os.path.exists(output_folder_name):
+        os.makedirs(output_folder_name)
+
+    sequence_df, binary_df = parse_vcf(vcf_file_path)
+    groups_dict = find_clusters(binary_df, min_group_size)
```
Comment (peterk87) [inline biohansel/cli.py]: What about outputting all the intermediate results to the output directory? Those files may be useful to the user so they can see how the clusters were determined. 

It would also be useful to create clusters at various group size thresholds. We could take a range of values from the command-line, e.g. . `--min-group-size 2-10` and within the output directory you could have subdirectories containing the schemes created at each distinct `min_group_size`

File: biohansel/cli.py [hunk 19]
```diff
@@ -276,6 +312,37 @@ def create(vcf_file_path, reference_genome_path, phylo_tree_path, distance_thres
...
+    if phylo_tree_path is not None:
+        new_tree=display_tree(phylo_tree_path, groups_dict)
```
Comment (peterk87) [inline biohansel/cli.py]: What about the return value from `display_tree`?

File: biohansel/cli.py [hunk 20]
```diff
@@ -247,26 +253,56 @@ def parse_comma_delimited_floats(ctx: click.Context, param: click.Option, value:
...
+              help='A unique name for the schema file that is generated, the default is just'
+                   '{bio_hansel-schema-reference_genome_name}-{schema_version}')
+@click.option('-m', '--schema-version',
+              required=False,
+              type=str,
+              help='An optional version number for the schema file that is generated')
+@click.option('-p', '--padding-sequence-length',
+              required=True,
+              type=int,
+              help='Output folder name in which schema file would be located'
```
[the remark this answers is not in the document]
Comment (gcttong) [reply biohansel/cli.py]: yes this has been addressed in commit 0175d925d1a6a25cbd9b292422c36869fe85914f

File: biohansel/cli.py [hunk 21]
```diff
@@ -247,26 +253,56 @@ def parse_comma_delimited_floats(ctx: click.Context, param: click.Option, value:
...
+              )
+@click.option('-f', '--reference-genome-format',
```
[the remark this answers is not in the document]
Comment (gcttong) [reply biohansel/cli.py]: yes this has been addressed in commit 0175d925d1a6a25cbd9b292422c36869fe85914f, I kept this variable since the genbank format definitely comes up alot as well. Also Biopython has a built in fasta parser, so that was the one I wanted to keep using. Let me know if that's ok

File: biohansel/cli.py [hunk 22]
```diff
@@ -276,6 +312,37 @@ def create(vcf_file_path, reference_genome_path, phylo_tree_path, distance_thres
...
+    click.secho(f'Output folder name: {output_folder_name}', fg='magenta')
+    click.secho(f'Scheme name: {schema_name}', fg='cyan')
+    click.secho(f'Reference genome format: {reference_genome_format}', fg='green')
+    click.secho(f'Min group size: {min_group_size}', fg='yellow')
+    click.secho(f'Padding Sequence length: {padding_sequence_length}', fg='blue')
     logging.info(f'Creating biohansel subtyping scheme from SNVs in "{vcf_file_path}" using reference genome '
                  f'"{reference_genome_path}" at {distance_thresholds if distance_thresholds else "all possible"} '
                  f'distance threshold levels.')
+    reference_genome_name = os.path.split(reference_genome_path)[-1]
+    reference_genome_name = reference_genome_name.split(".")[-2]
```
[the remark this answers is not in the document]
Comment (gcttong) [reply biohansel/cli.py]: yes this has been fixed in 181b812e4cf2e37bcb56db027de3ae454ef68697

File: biohansel/cli.py [hunk 23]
```diff
@@ -276,6 +312,37 @@ def create(vcf_file_path, reference_genome_path, phylo_tree_path, distance_thres
     click.secho(f'Reference genome file path: {reference_genome_path}', fg='red')
     click.secho(f'Phylogenetic tree file path: {phylo_tree_path}', fg='yellow')
     click.secho(f'Distance thresholds: {distance_thresholds}', fg='blue')
+    click.secho(f'Output folder name: {output_folder_name}', fg='magenta')
```
[the remark this answers is not in the document]
Comment (gcttong) [reply biohansel/cli.py]: yes this has been fixed in 181b812e4cf2e37bcb56db027de3ae454ef68697

File: biohansel/cli.py [hunk 24]
```diff
@@ -159,7 +163,7 @@ def subtype(scheme,
     input_contigs, input_reads = collect_inputs(**locals())
     if len(input_contigs) == 0 and len(input_reads) == 0:
         no_files_exception = click.UsageError('No input files specified!')
-        click.secho('Please see -h/--help for more info', err=True)
+        logging.info('Please see -h/--help for more info', err=True)
```
Comment (peterk87) [inline biohansel/cli.py]: Please revert this change. If there are issues with the code related to the `subtype` command, let me know in https://github.com/phac-nml/biohansel/pull/52

File: biohansel/cli.py [hunk 25]
```diff
@@ -247,35 +251,91 @@ def parse_comma_delimited_floats(ctx: click.Context, param: click.Option, value:
...
+              help='Length of sequence to be added around each SNV in the schema file,'
+              'if an even integer is provided, then the tile length would be n+1'
+              )
+@click.option('-f', '--reference-genome-format',
+              required=True,
+              type=click.Choice(['fasta', 'genbank']),
+              help='Reference genome file format: can be either fasta or genbank format'
+              )
+@click.option('-g', '--min-group-size',
+              type=click.Choice(['2', '3', '4', '5', '6', '7', '8', '9', '10']),
```
Comment (peterk87) [inline biohansel/cli.py]: Please see [my comment](https://github.com/phac-nml/biohansel/pull/58#discussion_r209313722) about making this option accepting a range of integer values. 

> Let's also set a default for this value to reduce the amount of things users need to specify via the command-line and that make sense like integers between 2 and 10 (or half of number of input samples) inclusive.

e.g. `hansel create ... -g 2-50` -> creates schemes with min group size from 2 to 50 inclusive.

File: biohansel/cli.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (peterk87) [reply biohansel/cli.py]: Sure, that's fine. However, please make this option optional (i.e. `required=False`) and try to find out the format either from the filename or try parsing using either format's BioPython parser. Let's try to let users run `biohansel create ...` while specifying as few options as possible where the defaults make sense given their input data. 

If a GenBank format file is supplied and it has annotations (e.g. Prokka output or RefSeq GenBank file), it may be useful to output what the markers target in a table (e.g. marker at position 123 in contig *abc* in genome *X* lies in CDS with gene name *xyz* and product "some protein"). This could be useful information for understanding the scheme output.

File: biohansel/cli.py (hunk 24 above)
Comment (gcttong) [reply biohansel/cli.py]: yes this has been addressed in d141a97fd0744cfd2f291e96d02ced7710935307

File: biohansel/cli.py (hunk 25 above)
Comment (gcttong) [reply biohansel/cli.py]: ok yes this has been addressed in d141a97fd0744cfd2f291e96d02ced7710935307

File: biohansel/cli.py [hunk 26]
```diff
@@ -247,26 +253,56 @@ def parse_comma_delimited_floats(ctx: click.Context, param: click.Option, value:
...
+              required=True,
+              type=int,
+              help='Output folder name in which schema file would be located'
+              )
+@click.option('-f', '--reference-genome-format',
+              required=True,
+              type=str,
+              help='Reference genome file format, i.e. fasta, genbank'
+              )
+@click.option('-t', '--min-group-size',
```
[the remark this answers is not in the document]
Comment (gcttong) [reply biohansel/cli.py]: yes this has been addressed in commit d141a97fd0744cfd2f291e96d02ced7710935307

File: biohansel/cli.py (hunk 18 above)
Comment (gcttong) [reply biohansel/cli.py]: yes this has been addressed in d141a97fd0744cfd2f291e96d02ced7710935307

File: biohansel/create/cluster_generator.py (hunk 16 above)
Comment (gcttong) [reply biohansel/create/cluster_generator.py]: ok yes this has been addressed in d141a97fd0744cfd2f291e96d02ced7710935307

File: biohansel/create/cluster_generator.py (hunk 17 above)
Comment (gcttong) [reply biohansel/create/cluster_generator.py]: ok yes this has been addressed in d141a97fd0744cfd2f291e96d02ced7710935307

File: biohansel/create/display_tree.py [hunk 27]
```diff
@@ -0,0 +1,27 @@
+import logging
+
+from ete3 import Tree
+
+from typing import Dict
+
+
+def display_tree(phylo_tree_path: str, groups_dict: Dict[str, str]) -> str:
```
[the remark this answers is not in the document]
Comment (gcttong) [reply biohansel/create/display_tree.py]: yes this has been addressed in commit b9b885770cc293b913fca07f7f2e79cb31242e75

File: biohansel/create/schema_generator.py (hunk 1 above)
Comment (gcttong) [reply biohansel/create/schema_generator.py]: yes this has been addressed in 46cbc1f527bbf6d0225c3fdb51585fbe19ffc4f7

File: biohansel/cli.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (gcttong) [reply biohansel/cli.py]: yes this has been addressed in commit b554b060efeb93d1eac0414a82f84a6088ac8b6c where the user has the option to specify the format, but is not required to

File: biohansel/create/io/output.py (hunk 2 above)
Comment (gcttong) [reply biohansel/create/io/output.py]: yes this was addressed in b3479441f774fa2b3fc3f01a9a8e74656a4aaaeb

File: biohansel/create/io/output.py (hunk 3 above)
Comment (gcttong) [reply biohansel/create/io/output.py]: yes this was addressed in b3479441f774fa2b3fc3f01a9a8e74656a4aaaeb

File: biohansel/create/schema_generator.py (hunk 5 above)
Comment (gcttong) [reply biohansel/create/schema_generator.py]: yes this has been addressed in 784a18e5cc8d182b8bce41777bb9a7b0547aaf04

File: biohansel/create/cluster_generator.py (hunk 6 above)
Comment (gcttong) [reply biohansel/create/cluster_generator.py]: I kept this as a function because the values would then get passed to row_subtype after

File: biohansel/create/cluster_generator.py (hunk 7 above)
Comment (gcttong) [reply biohansel/create/cluster_generator.py]: addressed in 99966c049e443fa7a170738ee145b3be2d7f64d4

File: biohansel/create/cluster_generator.py (hunk 9 above)
Comment (gcttong) [reply biohansel/create/cluster_generator.py]: addressed in 99966c049e443fa7a170738ee145b3be2d7f64d4

File: biohansel/create/cluster_generator.py (hunk 10 above)
Comment (gcttong) [reply biohansel/create/cluster_generator.py]: addressed in 99966c049e443fa7a170738ee145b3be2d7f64d4

File: biohansel/create/cluster_generator.py (hunk 8 above)
Comment (gcttong) [reply biohansel/create/cluster_generator.py]: addressed in b5b061a66ef9372537e2fcb8891f2f7a747b98db

File: biohansel/create/cluster_generator.py (hunk 15 above)
Comment (gcttong) [reply biohansel/create/cluster_generator.py]: yes this has been addressed in 94b2c0213878ff804e065eb970dc3023030acfcc

File: biohansel/create/cluster_generator.py (hunk 14 above)
Comment (gcttong) [reply biohansel/create/cluster_generator.py]: yes this has been addressed in 94b2c0213878ff804e065eb970dc3023030acfcc

File: biohansel/create/cluster_generator.py (hunk 13 above)
Comment (gcttong) [reply biohansel/create/cluster_generator.py]: yes this has been addressed in 94b2c0213878ff804e065eb970dc3023030acfcc

File: biohansel/create/cluster_generator.py (hunk 12 above)
Comment (gcttong) [reply biohansel/create/cluster_generator.py]: yes this has been addressed in 94b2c0213878ff804e065eb970dc3023030acfcc

File: biohansel/create/cluster_generator.py (hunk 11 above)
Comment (gcttong) [reply biohansel/create/cluster_generator.py]: yes this has been addressed in 94b2c0213878ff804e065eb970dc3023030acfcc

**Comment by glabbe** [comment]
We are working on a separate repo for the scheme creation tool: BioCanon <https://github.com/phac-nml/bioCanon>
